ADR: Mailbox API visibility and usage (#3333)
linagora/james-project: src/adr/0034-mailbox-api-visibility-and-usage.md

# 34. Mailbox API visibility and usage

Date: 2020-04-27

## Status

Accepted (lazy consensus)

## Context

> For some reasons, all mailboxes implementations rely on some abstract classes located in `mailbox-store` module (like 

Reviewer: Can you rephrase the sentence to not start it by `For some (unexplained) reasons,`

From a reviewer perspective, it avoids asking the question (Why?). Removing the prefix is ok to me.

> For some reasons, all mailboxes implementations rely on some abstract classes located in `mailbox-store` module (like 

Reviewer: mailbox/store defines some common tools to implement mailbox/api. Speaking of abstract classes is maybe a bit too much technical still.

`StoreMailboxManager` and the `Mapper`s classes). For example, a `CassandraMailboxmanager` has to extend `StoreMailboxManager` 
(that implements `Mailboxmanager` from the `mailbox-api`) that requires the implementation of some `Mapper`s.

`Mapper`s are designed to provide low-level functions and methods on mailboxes. It's recurrent that we are tempted in 
James, outside of the `mailbox` modules, to rely on some of those classes located in `mailbox-store` to have an easier 
access on some user's mailboxes or messages. 

> 

Reviewer: Is there some services handled at the manager layer that are not covered by the mapper layer?

Reviewer: The manager layer do further validation including right checking, event dispatching (resulting in message search index indexing, current quota calculation mainly), quota validation. Not relying on the manager layer is thus error prone and can lead to security vulnerabilities.

## Decision

We should never rely on classes defined in `mailbox-store` outside of the `mailbox` modules. The right way would be to 
always rely on the `Manager`s defined in `mailbox-api` module to access mailboxes and messages.

We should ensure the correctness of `Manager`s implementations by providing contract tests and not by sharing abstract 
classes.

## Consequences

> The issue with this design is that in this day, we can't seem to be able to act on a single message without knowing a 

Reviewer 2: Sorry but I don't understand what it means.
Manager layer is responsible to ensuring access control on entities it manages.
A Message should be accessed only by either a User having access to it or to a technical component that has specific rights.
How is this problem related to the design describe previously?
Are you referring to the fact that by-passing access control is very handy? Because it's not stated explicitly in the Context.

Author: That was the all problem of this PR, starting the discussion: https://github.com/linagora/james-project/pull/3035, where I had to use the MailboxMapper outside of mailbox modules to be able to recompute the fast view projection of a single message of an user, as there was no other way at the moment. 

The conclusion, if I'm not wrong, was to bypass indeed the access control with the session, which would require a refactoring of this one. As the session refactoring should follow in an other ADR, I was not sure how much I should tell in this one. 

You are saying the context should state more that sometimes we need to bypass that access-control to access a message elsewhere?

Reviewer: Currently, right enforcement requires a session to be open with the owner username for interacting with the manager layer, which is unknown given a simple `mailboxId`.

Reviewer 2: Yes, I'm saying that, even if previous discussion is enough for me to know what you mean, it deserves to be written in Context part.

* `mailbox-adapter` because `Authenticator` and `Authorizator` are part of the `mailbox-store`

The manager layer do further validation including right checking, event dispatching (resulting in message search index 
indexing, current quota calculation mainly), quota validation. Not relying on the manager layer is thus error prone 
and can lead to security vulnerabilities.

## Decision

We should never rely on classes defined in `mailbox-store` outside of the `mailbox` modules (except on some cases 
limited to the test scope). The right way would be to always rely on the `Manager`s defined in `mailbox-api` module to 
access mailboxes and messages, as the `mailbox-api` module defines the API on how to use a mailbox.

We should ensure the correctness of `Manager`s implementations by providing contract tests and not by sharing abstract 
classes.

Regarding the modules wrongly relying already on `mailbox-store`, we can:

* `mailbox-adapter`: move `Authenticator` and `Authorizator` to `mailbox-api`

## Consequences

We need to introduce some refactorings to be able to rely fully on `mailbox-api` in new emerging cases. For example, 

> our `mailbox-api` still can't seem to be able to act on a single message without knowing a user it belongs to. It 

Reviewer 2: ```suggestion
our `mailbox-api` still lacks APIs to handle messages by their MessageId. It 
```

## Context

All mailboxes implementations rely on `mailbox-store` module that defines some common tools to implement the `mailbox-api`
(representing the API defining how to use a mailbox). 

For example, a `CassandraMailboxmanager` has to extend `StoreMailboxManager` (that implements `Mailboxmanager` from the 
`mailbox-api`) that requires the implementation of some `Mapper`s.

`Mapper`s are designed to provide low-level functions and methods on mailboxes. It's recurrent that we are tempted in 
James, outside of the `mailbox` modules, to rely on some of those common tools located in `mailbox-store` to have an 
easier access on some user's mailboxes or messages. 

Like for example, using a `Mapper` outside to be able to retrieve a message with only its `MessageId`, which is not 
currently possible at the `Manager`'s level, which tends to violate `mailbox-api`'s role and primary mission.

As a matter of fact, we have currently such uses of `mailbox-store` in James:

* `mailbox-adapter` because `Authenticator` and `Authorizator` are part of the `mailbox-store`
* `mpt-imapmailbox-jpa-lucene` for tests

> * `mpt-imapmailbox-elasticsearch` for tests

Reviewer: mpt projects are tests, they are okish to me.

## Decision

We should never rely on classes defined in `mailbox-store` outside of the `mailbox` modules (except on some cases 
limited to the test scope). The right way would be to always rely on the `Manager`s defined in `mailbox-api` module to 
access mailboxes and messages, as the `mailbox-api` module defines the API on how to use a mailbox.

We should ensure the correctness of `Manager`s implementations by providing contract tests and not by sharing abstract 
classes.

Regarding the modules wrongly relying already on `mailbox-store`, we can:

* `mailbox-adapter`: move `Authenticator` and `Authorizator` to `mailbox-api`
* `mpt-imapmailbox-jpa-lucene`: limit it to the test scope

> * `mpt-imapmailbox-elasticsearch`: limit it to the test scope

Reviewer: For MPT it's not needed IMO as they are testing projects.

## Consequences

> We need to introduce some refactorings to be able to rely fully on `mailbox-api` in new emerging  cases. For example, 

Reviewer 3: ```suggestion
We need to introduce some refactorings to be able to rely fully on `mailbox-api` in new emerging cases. For example, 
```
